src/App.js

```diff
@@ -21,45 +23,106 @@ const generateSquares = () => {
+  const [player, setPlayer] = useState(playerOne);
+  const [isGameOver, setGameStatus] = useState(false);
+  const [winner, setWinner] = useState(winnerInitial);
 
-  // Wave 2
-  // You will need to create a method to change the square
-  //   When it is clicked on.
-  //   Then pass it into the squares as a callback
+  const updateOnClick = (updatedSquare) => {
+    const squaresData = squares.map((row) => {
+      row.map((square) => {
+        if (square.id === updatedSquare.id) {
+          if (square.value === '') {
+            square.value = player;
+            const updatedWinner = checkForWinner();
+            if (updatedWinner) {
+              setWinner(updatedWinner);
+              setGameStatus(true);
+            }
+            setPlayer(player === playerOne ? playerTwo : playerOne);
+          }
+        }
+      });
+    });
+  };
+
+  const renderPlayerTurn = () => {
+    let playerTurn = '';
+    if (winner === winnerInitial) {
+      playerTurn = `It's ${player}'s turn!`;
+    } else {
+      playerTurn = `🎉 Way to go ${winner} 🥳`;
+    }
+    return playerTurn;
+  };
 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if
-    //    all three squares have the same value.
+    let i = 0;
+    // Check all the rows and columns for a winner
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    // Check Top-Left to bottom-right diagonal
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
+
+    // Check Top-right to bottom-left diagonal
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+    return '';
   };
 
   const resetGame = () => {
-    // Complete in Wave 4
+    setSquares(generateSquares());
+    setWinner(winnerInitial);
+    setGameStatus(false);
+    setPlayer(playerOne);
   };
 
   return (
     <div className="App">
       <header className="App-header">
         <h1>React Tic Tac Toe</h1>
-        <h2>The winner is ... -- Fill in for wave 3 </h2>
-        <button>Reset Game</button>
+        <h2>Winner is {winner}</h2>
```

Comment:
Another option here would be to hide the entire winner message with conditional rendering until you have an actual winner.